File: src/jdwp/events.ts
// Decode the body of an Event.Composite packet. One composite carries a suspend
// policy and a list of sub-events, each tagged by event kind with a kind-shaped
// payload. Only the kinds cappu requests are decoded; unknown kinds stop the
// scan (their length is not self-describing, so we cannot skip them safely).
//
// Port reference for togo/internal/jdwp/events.go.

import { ByteReader, type IdSizes } from "./idCodec.ts";
import { type Location, readLocation } from "./commands.ts";
import { EventKind } from "./protocol.ts";

export type JdwpEvent =
  | { kind: typeof EventKind.VM_START; requestId: number; thread: bigint }
  | { kind: typeof EventKind.VM_DEATH; requestId: number }
  | { kind: typeof EventKind.THREAD_START; requestId: number; thread: bigint }
  | { kind: typeof EventKind.THREAD_DEATH; requestId: number; thread: bigint }
  | { kind: typeof EventKind.BREAKPOINT; requestId: number; thread: bigint; location: Location }
  | { kind: typeof EventKind.SINGLE_STEP; requestId: number; thread: bigint; location: Location }
  | {
      kind: typeof EventKind.CLASS_PREPARE;
      requestId: number;
      thread: bigint;
      refTypeTag: number;
      typeId: bigint;
      signature: string;
      status: number;
    };

export interface Composite {
  suspendPolicy: number;
  events: JdwpEvent[];
}

export function decodeComposite(data: Buffer, sizes: IdSizes): Composite {
  const r = new ByteReader(data);
  const suspendPolicy = r.u1();
  const count = r.u4();
  const events: JdwpEvent[] = [];
  for (let i = 0; i < count; i++) {
    const kind = r.u1();
    switch (kind) {
      case EventKind.VM_START:
        events.push({ kind, requestId: r.i4(), thread: r.id(sizes.objectID) });
        break;
      case EventKind.VM_DEATH:
        events.push({ kind, requestId: r.i4() });
        break;
      case EventKind.THREAD_START:
      case EventKind.THREAD_DEATH:
        events.push({ kind, requestId: r.i4(), thread: r.id(sizes.objectID) });
        break;
      case EventKind.BREAKPOINT:
      case EventKind.SINGLE_STEP:
        events.push({
          kind,
          requestId: r.i4(),
          thread: r.id(sizes.objectID),
          location: readLocation(r, sizes),
        });
        break;
      case EventKind.CLASS_PREPARE:
        events.push({
          kind,
          requestId: r.i4(),
          thread: r.id(sizes.objectID),
          refTypeTag: r.u1(),
          typeId: r.id(sizes.referenceTypeID),
          signature: r.string(),
          status: r.i4(),
        });
        break;
      default:
        return { suspendPolicy, events }; // unknown kind: cannot skip, stop here
    }
  }
  return { suspendPolicy, events };
}

File: togo/internal/jdwp/events.go
package jdwp

// Decode the body of an Event.Composite packet: a suspend policy and a list of
// sub-events, each tagged by event kind with a kind-shaped payload. Only the
// kinds cappu requests are decoded; an unknown kind stops the scan (its length
// is not self-describing). Port of src/jdwp/events.ts.

// Event is one decoded JDWP sub-event. Fields are populated per Kind.
type Event struct {
	Kind       byte
	RequestID  int32
	Thread     uint64
	Location   Location
	RefTypeTag byte
	TypeID     uint64
	Signature  string
	Status     int32
}

// Composite is a decoded Event.Composite body.
type Composite struct {
	SuspendPolicy byte
	Events        []Event
}

func DecodeComposite(data []byte, s IDSizes) Composite {
	r := NewReader(data)
	comp := Composite{SuspendPolicy: r.U1()}
	count := int(r.U4())
	for range count {
		kind := r.U1()
		switch kind {
		case EKVMStart:
			comp.Events = append(comp.Events, Event{Kind: kind, RequestID: r.I4(), Thread: r.ID(s.ObjectID)})
		case EKVMDeath:
			comp.Events = append(comp.Events, Event{Kind: kind, RequestID: r.I4()})
		case EKThreadStart, EKThreadDeath:
			comp.Events = append(comp.Events, Event{Kind: kind, RequestID: r.I4(), Thread: r.ID(s.ObjectID)})
		case EKBreakpoint, EKSingleStep:
			comp.Events = append(comp.Events, Event{
				Kind:      kind,
				RequestID: r.I4(),
				Thread:    r.ID(s.ObjectID),
				Location:  readLocation(r, s),
			})
		case EKClassPrepare:
			comp.Events = append(comp.Events, Event{
				Kind:       kind,
				RequestID:  r.I4(),
				Thread:     r.ID(s.ObjectID),
				RefTypeTag: r.U1(),
				TypeID:     r.ID(s.ReferenceTypeID),
				Signature:  r.String(),
				Status:     r.I4(),
			})
		default:
			return comp // unknown kind: cannot skip, stop here
		}
	}
	return comp
}
